front50-core/src/main/groovy/com/netflix/spinnaker/front50/model/pluginartifact/PluginArtifactService.java

```diff
@@ -0,0 +1,102 @@
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.netflix.spinnaker.front50.model.pluginartifact;
+
+import com.netflix.spinnaker.front50.exception.NotFoundException;
+import com.netflix.spinnaker.front50.validator.GenericValidationErrors;
+import com.netflix.spinnaker.front50.validator.PluginArtifactValidator;
+import com.netflix.spinnaker.kork.exceptions.UserException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import javax.annotation.Nonnull;
+import org.springframework.stereotype.Component;
+import org.springframework.validation.Errors;
+
+@Component
+public class PluginArtifactService {
+
+  private final PluginArtifactRepository repository;
+  private final List<PluginArtifactValidator> validators;
+
+  public PluginArtifactService(
+      PluginArtifactRepository repository, List<PluginArtifactValidator> validators) {
+    this.repository = repository;
+    this.validators = validators;
+  }
+
+  public Collection<PluginArtifact> findAll() {
+    return repository.all();
+  }
+
+  public Collection<PluginArtifact> findAllByService(@Nonnull String service) {
+    return repository.getByService(service);
+  }
+
+  public PluginArtifact upsert(@Nonnull PluginArtifact pluginArtifact) {
+    validate(pluginArtifact);
+
+    try {
+      repository.findById(pluginArtifact.getId());
+      repository.update(pluginArtifact.getId(), pluginArtifact);
+      return pluginArtifact;
+    } catch (NotFoundException e) {
+      return repository.create(pluginArtifact.getId(), pluginArtifact);
+    }
+  }
+
+  public void delete(@Nonnull String id) {
+    repository.delete(id);
+  }
+
+  public PluginArtifact createRelease(@Nonnull String id, @Nonnull PluginArtifact.Release release) {
+    PluginArtifact artifact = repository.findById(id);
+
+    artifact.getReleases().add(release);
+
+    return upsert(artifact);
+  }
+
+  public PluginArtifact deleteRelease(@Nonnull String id, @Nonnull String releaseVersion) {
+    PluginArtifact artifact = repository.findById(id);
+
+    new ArrayList<>(artifact.getReleases())
+        .forEach(
+            release -> {
+              if (release.getVersion().equals(releaseVersion)) {
+                artifact.getReleases().remove(release);
+              }
+            });
+
+    return upsert(artifact);
+  }
+
+  private void validate(PluginArtifact pluginArtifact) {
+    Errors errors = new GenericValidationErrors(pluginArtifact);
+    validators.forEach(v -> v.validate(pluginArtifact, errors));
+    if (errors.hasErrors()) {
+      throw new ValidationException(errors);
+    }
+  }
+
```

Comment:
Do we need  exception handler serializing this exception or we have a generic exception handler somewhere in Kork  ?

Reply:
We have a generic handler: https://github.com/spinnaker/kork/blob/master/kork-web/src/main/java/com/netflix/spinnaker/kork/web/exceptions/GenericExceptionHandlers.java